### ISSUE TITLE
Principals in aws_s3_bucket_policy should be ordered for plan

### ISSUE THREAD

**Issue by sblask** [hashicorp/terraform-provider-aws#18878]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Principals in a bucket policy are not ordered so when principals are added/removed, the diff in the plan shows not only the addition/removal, but also the shuffling of principals. Example: this is the diff I get when adding "arn:aws:iam::XXXXXXXXXXXXX:role/user-d" and "arn:aws:iam::XXXXXXXXXXXXX:role/instance-d"

```diff
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/instance-a",
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/instance-b",
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/instance-c",
                              - "arn:aws:iam::XXXXXXXXXXXXX:role/user-b",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/user-d",
                                "arn:aws:iam::XXXXXXXXXXXXX:role/user-c",
                                "arn:aws:iam::XXXXXXXXXXXXX:role/user-a",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/user-b",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-d",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-c",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-a",
                              + "arn:aws:iam::XXXXXXXXXXXXX:role/instance-b",
                            ]
                        }
```

Ideally I should only see 

```diff
...
+ "arn:aws:iam::XXXXXXXXXXXXX:role/instance-d"
+ "arn:aws:iam::XXXXXXXXXXXXX:role/user-d"
...
```

### New or Affected Resource(s)

* aws_s3_bucket_policy

**Comment by YakDriver** [comment]
@sblask This is another interesting one! There are a few pieces at play here. `aws_s3_bucket_policy` uses https://github.com/jen20/awspolicyequivalence to check if the old and new policies are equivalent. This is a simple yes/no. Here, we all agree that the policies are not equivalent because something has been added. So, up to that point, the AWS provider is doing the right thing, I think.

There are 2 more pieces to the puzzle:

1. **the diff the provider displays:** the Terraform Plugin SDK figures out what the differences are between old and new values. It will provide these to a resource if asked, and often resources use the information to make targeted updates. You can see [an example of this with tags for S3 bucket](https://github.com/hashicorp/terraform-provider-aws/blob/9a609f3305378574b57e5f1fbf695a38d66bf4b2/aws/resource_aws_s3_bucket.go#L711-L722). However, the `aws_s3_bucket_policy` doesn't use the diff so what you're seeing, doesn't correlate to 2 below.
2. **what actually happens vis-à-vis AWS itself:** `aws_s3_bucket_policy` does not have an Update and instead sends the new policy whole to AWS to replace the existing policy in its entirety. This is a limitation of AWS itself. There's no way to update a bucket's policy. You can only ["Put"](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.PutBucketPolicy) one, which replaces the old one.

That's a long way of saying that there's not anything we can actually do as this relates to an inherent limitation of AWS. If AWS maintained a list (or set) of principal relative to an S3 bucket, we could surgically update. However, since everything is wrapped in an IAM policy, and it's all or nothing, our hands are tied.

**Comment by gdavison** [comment]
This can be addressed in the provider, though it may need an upstream change to https://github.com/jen20/awspolicyequivalence

**Comment by YakDriver** [comment]
This should be fixed by #21997
